test/forward/forward.jl

```diff
@@ -20,22 +29,22 @@ end
 
 @test D(x -> pow(x, 3), 2) == 12
 
-@test D(1) do x
-  f(y) = x = x*y
-  D(f, 1)
-  D(f, 1)
-end == 1
-
-@test D(x -> D(y -> x = y, x)*x, 1) == 1
-
-@test D(1) do x
-  D(2) do y
-    D(3) do z
-      x = z * y
-    end
-  end
-  x
-end == 0
+# @test D(1) do x
+#   f(y) = x = x*y
+#   D(f, 1)
+#   D(f, 1)
+# end == 1
+
+# @test D(x -> D(y -> x = y, x)*x, 1) == 1
+
+# @test D(1) do x
+#   D(2) do y
+#     D(3) do z
+#       x = z * y
+#     end
+#   end
+#   x
+# end == 0
 
 @test D(x -> abs(x+2im), 1) == gradient(x -> abs(x+2im), 1)[1]
```

Comment:
This doesn't currently pass. I'm pretty sure that the `Forward` implementation is doing the correct thing, and `gradient` is wrong. `D` (correctly) returns a `Real` as the derivative w.r.t `x`, while `gradient` returns a `Complex`. @sethaxen do you agree that Zygote is doing the wrong thing here?

Reply:
ChainRulesCore has an issue open about this: https://github.com/JuliaDiff/ChainRulesCore.jl/issues/176. It's also related to https://github.com/JuliaDiff/ChainRules.jl/issues/232. The issue is that our rules for addition between real and complex (`+(x, 2im)`) will pull back a complex adjoint to `x`. 
So we'd expect these to do the same thing, but
```julia
julia> gradient(x -> abs(x+2im), 1)[1]
0.4472135954999579 + 0.8944271909999159im

julia> gradient(x -> abs(complex(x, 2)), 1)[1]
0.4472135954999579
```
The first issue above links to some Zygote discussions. Essentially, Zygote treats all reals as actually complex, so from that perspective it is doing the right thing. I just don't think that's the right way to go.

Reply:
Ahh interesting. It's interesting that `Zygote.Forward` doesn't naturally do the same thing as `Zygote`.

I guess that the question is what the right way forward for us is with `Forward`. Mike previously had some extra code for `abs` to make it do the Zygote-y thing.

I wonder whether it's acceptable to utilise a different convention in `Forward` than what `Zygote` currently does...

@DhairyaLGandhi what are your thoughts on the way that Zygote currently handles complex numbers / have you had a chance to review the various ChainRules discussions about complex numbers?